src/biogeophys/SnowSnicarMod.F90

```diff
@@ -578,73 +774,158 @@ subroutine SNICAR_RT (flg_snw_ice, bounds, num_nourbanc, filter_nourbanc,  &
+                  ! Define local Mie parameters based on snow grain size and aerosol species retrieved from a lookup table.
+
+                  ! Spherical snow: single-scatter albedo, mass extinction coefficient, asymmetry factor
                   if (flg_slr_in == 1) then
                      do i=snl_top,snl_btm,1
                         rds_idx = snw_rds_lcl(i) - snw_rds_min_tbl + 1
                         ! snow optical properties (direct radiation)
                         ss_alb_snw_lcl(i)      = ss_alb_snw_drc(rds_idx,bnd_idx)
-                        asm_prm_snw_lcl(i)     = asm_prm_snw_drc(rds_idx,bnd_idx)
                         ext_cff_mss_snw_lcl(i) = ext_cff_mss_snw_drc(rds_idx,bnd_idx)
+                        if (sno_shp(i) == 'sphere') asm_prm_snw_lcl(i) = asm_prm_snw_drc(rds_idx,bnd_idx)
                      enddo
                   elseif (flg_slr_in == 2) then
                      do i=snl_top,snl_btm,1
                         rds_idx = snw_rds_lcl(i) - snw_rds_min_tbl + 1
                         ! snow optical properties (diffuse radiation)
                         ss_alb_snw_lcl(i)      = ss_alb_snw_dfs(rds_idx,bnd_idx)
-                        asm_prm_snw_lcl(i)     = asm_prm_snw_dfs(rds_idx,bnd_idx)
                         ext_cff_mss_snw_lcl(i) = ext_cff_mss_snw_dfs(rds_idx,bnd_idx)
+                        if (sno_shp(i) == 'sphere') asm_prm_snw_lcl(i) = asm_prm_snw_dfs(rds_idx,bnd_idx)
                      enddo
                   endif
 
-                  ! aerosol species 1 optical properties
-                  ss_alb_aer_lcl(1)        = ss_alb_bc1(bnd_idx)      
-                  asm_prm_aer_lcl(1)       = asm_prm_bc1(bnd_idx)
-                  ext_cff_mss_aer_lcl(1)   = ext_cff_mss_bc1(bnd_idx)
+                  ! Nonspherical snow: shape-dependent asymmetry factors
+                  do i=snl_top,snl_btm,1
+
+                     select case (sno_shp(i))
+                     case ('spheroid')
+                        diam_ice = 2._r8 * snw_rds_lcl(i)   ! unit: microns
+                        if (sno_fs(i) == 0._r8) then
+                           fs_sphd = 0.929_r8  ! default; He et al. (2017), Table 1
+                        else
+                           fs_sphd = sno_fs(i) ! user specified value
+                        endif
+                        fs_hex = 0.788_r8      ! reference shape factor
+                        if (sno_AR(i) == 0._r8) then
+                           AR_tmp = 0.5_r8     ! default; He et al. (2017), Table 1
+                        else
+                           AR_tmp = sno_AR(i)  ! user specified value
+                        endif
+                        do igb = 1,7
+                           g_ice_Cg_tmp(igb) = g_b0(igb) * ((fs_sphd/fs_hex)**g_b1(igb)) * (diam_ice**g_b2(igb))   ! Eq.7, He et al. (2017)
+                           gg_ice_F07_tmp(igb) = g_F07_c0(igb) + g_F07_c1(igb)*AR_tmp + g_F07_c2(igb)*(AR_tmp**2._r8)  ! Eqn. 3.1 in Fu (2007)
```

Comment:
Also, looking at the code for examples like this where I can improve computational efficiency by replacing a**2 with a*a and a/2 with a*0.5

Reply:
I hope I got most (if not all) of those. 

I also added some `._r8` to numbers that needed to be floats.